crates/next-core/src/app_source.rs

```diff
@@ -344,131 +349,149 @@ async fn create_app_source_for_directory(
     let mut sources = Vec::new();
     let mut page = None;
     let mut files = HashMap::new();
-    if let DirectoryContent::Entries(entries) = &*input_dir.read_dir().await? {
-        for (name, entry) in entries.iter() {
-            if let &DirectoryEntry::File(file) = entry {
-                if let Some((name, _)) = name.rsplit_once('.') {
-                    match name {
-                        "page" => {
-                            page = Some(file);
-                        }
-                        "layout" | "error" | "loading" | "template" | "not-found" | "head" => {
-                            files.insert(name.to_string(), file);
-                        }
-                        _ => {
-                            // Any other file is ignored
-                        }
+
+    let DirectoryContent::Entries(entries) = &*input_dir.read_dir().await? else {
+        return Ok(CombinedContentSource { sources }.cell());
+    };
+
+    let allowed_extensions = &*page_extensions.await?;
+
+    for (name, entry) in entries.iter() {
+        if let &DirectoryEntry::File(file) = entry {
+            if let Some((name, ext)) = name.rsplit_once('.') {
+                if !allowed_extensions.iter().any(|allowed| allowed == ext) {
```

Comment:
Should this perhaps only apply to `page`/`error`/`loading`?

Reply:
I don't think so, why?

Reply:
It feels weird that `pageExtensions` would apply to things that aren't technically pages, like `head` or `layout`, but I guess it makes sense to support these too, so 🤷

Reply:
Well, we're just copying next.js behaviour

It treats all of them the same as far as I can tell https://github.com/vercel/next.js/blob/canary/packages/next/src/build/webpack/loaders/next-app-loader.ts#L12-L12